proposals/xxxx-package-manager-manifest-api-redesign.md

```diff
@@ -0,0 +1,674 @@
+    specifying multiple system packages by changing the payload to be an array.
+
+    <details>
+      <summary>View diff and example</summary>
+      <p>
+
+     Diff:
+    ```diff
+    enum SystemPackageProvider {
+    -    case Brew(String)
+    +    case brew([String])
+
+    -    case Apt(String)
+    +    case apt([String])
+    }
+    ```
+
+    Example:
+
+    ```diff
+    let package = Package(
+        name: "Copenssl",
+        pkgConfig: "openssl",
+        providers: [
+    -        .Brew("openssl"),
+    +        .brew(["openssl"]),
+
+    -        .Apt("openssl-dev"),
+    +        .apt(["openssl", "libssl-dev"]),
+        ]
+    )
+    ```
+    </p></details>
+
+
+* Remove implicit target dependency rule for test targets.
+
+    There is an implicit test target dependency rule: a test target "FooTests"
+    implicity depends on a target "Foo", if "Foo" exists and "FooTests" doesn't
+    explicitly declare any dependency. We propose to remove this rule because:
+
+    1. It is a non obvious "magic" rule that has to be learned.
+    2. It is not possible for "FooTests" to remove dependency on "Foo" while
+       having no other (target) dependency.
+    3. It makes real dependencies less discoverable.
+    4. It may cause issues when we get support for mechanically editing target
+       dependencies.
+
+* Use factory methods for creating objects.
+
+    We propose to always use factory methods to create objects except for the
+    main `Package` object. This gives a lot of flexibility and extensibility to
+    the APIs because Swift's type system can infer the top level type in a
+    context and allow using the shorthand dot syntax.
+
+    Concretely, we will make these changes:
+
+    * Add a factory method `target` to `Target` class and change the current
+      initializer to private.
+
+        <details>
+          <summary>View example and diff</summary>
+          <p>
+
+        Example:
+
+        ```diff
+        let package = Package(
+            name: "Foo",
+            target: [
+    -           Target(name: "Foo", dependencies: ["Utility"]),
+    +           .target(name: "Foo", dependencies: ["Utility"]),
+            ]
+        )
+        ```
+        </p></details>
+
+    * Introduce a `Product` class with two subclasses: `Executable` and
+      `Library`.  These subclasses will be nested inside `Product` class
+      instead of being a top level declaration in the module. Nesting will give
+      us a namespace for products and it is easy to find all the supported
+      products when the product types grows to a large number. We will add two
+      factory methods to `Product` class: `library` and `executable` to create
+      respective products.
+
+        ```swift
+        /// Represents a product.
+        class Product {
+        
+            /// The name of the product.
+            let name: String
+
+            /// The names of the targets in this product.
+            let targets: [String]
```

Comment:
This should be private, since not all products in the future will include targets and we don't want people to rely on it I don't think